### PR TITLE
Added balloon notifications with tray icon

### PR DESCRIPTION
**Pull request by Necktrox**
## Description

This pull requests aims to add notification functionality in the tray icon area on Windows task bar. It implements an unique class to manage creation, deletion of the tray icon and to show balloon notifications. The balloon notifications support 4 types of icons for the message and it's also possible to silence the notification sound.
## Features
- Tray icon
- Balloon notifications
- 4 notification icons (default, info, warning, error)
- Enable/disable notification sound
- Spam prevention
## Preview screenshots

Preview for tray notification
Preview for modified settings
## Notes

Any suggestions are welcome.

### CODE REVIEW

**Comment by AleksCore** [comment]
Wow! Nice job!

File: MTA10/mods/shared_logic/lua/CLuaFunctionDefs.Output.cpp [hunk 1]
```diff
@@ -227,7 +227,7 @@ int CLuaFunctionDefs::CreateTrayNotification ( lua_State* luaVM )
     bool useSound = true;
     CScriptArgReader argStream ( luaVM );
     argStream.ReadString ( strText );
-    argStream.ReadString ( strType, "noicon" );
+    argStream.ReadString ( strType, "default" );
```
Comment (qaisjp) [inline MTA10/mods/shared_logic/lua/CLuaFunctionDefs.Output.cpp]: Nice work.

By the way, this needs to be an enum. See the code for setCameraFieldOfView for an example.

Comment (Necktrox) [inline MTA10/mods/shared_logic/lua/CLuaFunctionDefs.Output.cpp]: Ok, no problem. I was looking at the enum code earlier, but I wasn't sure if that was required.
@qaisjp

**Comment by Necktrox** [comment]
Okay, the only problems I had was to find the correct location for the code parts, but it works.
@qaisjp

**Comment by qaisjp** [comment]
Nice work!

Please can you explain what the spam prevention is?

**Comment by Necktrox** [comment]
I prevent spam by blocking the function for a little period with GetTickCout64. Without this, a script could crash explorer, which happened on my computer.

**Comment by qaisjp** [comment]
Okies, just tested your PR.

First thing I notice is that the MTA icon pops up in my taskbar once I have joined a server. Would it be possible for you to make the icon only appear when notifications are sent?

You seem to already have code in there to destroy the tray icon when you leave a server but this doesn't work correctly (the icon does not get destroyed).

**Comment by Necktrox** [comment]
I will work on it in the next days, but IIRC there were issues with sending notifications right after creating the tray icon. Furthermore, the notification disappeared after destroying the tray icon, but I am not sure and I can't test it now.

**Comment by AlexTMjugador** [comment]
I think that adding a _title_ variable to the tray balloons class would be a nice touch. That way players would be able to know what is the topic of that message more easily, if the script wants to do so. However, it could be abused a bit to confuse players, but the way I see it formatting the title like "Multi Theft Auto: San Andreas — _CUSTOM THING HERE_" would be enough to avoid that.

I look forward to seeing this PR in MTA's trunk!

**Comment by Necktrox** [comment]
@AlexTMjugador I would keep it as it is. You could add a title or topic in the body section with enclosed brackets (e.g. "[News] ...") - @Jusonex proposed to remove the ability to change the title to avoid abuse.

**Comment by AlexTMjugador** [comment]
I understand, @Necktrox - that is a fair argument. It was just an idea I thought of when first watching this PR, to improve it even more, and I wanted to share it with the author :)

**Comment by Necktrox** [comment]
@qaisjp 
1) It is possible, BUT it won't be possible to show the MTA icon in the message AND the notification will disappear if you hover in the area, where the tray icon should have been actually.

2) Fixed in (both) last commits.

**Comment by qaisjp** [comment]
I agree that we should not have the ability to change the title.

Also, what if someone set the message to "Enter windows password to continue.."? Perhaps we should change the title to **"MTA Server says..."**?

**Comment by Necktrox** [comment]
It's not difficult to change the title, but I don't really see what this little change should avoid in practice. It won't stop server owners from showing a GUI input with similiar caption. Can you elaborate your argument?

**Comment by qaisjp** [comment]
Fair point. I just think that seeing a Windows notification that comes from MTA on a desktop looks more official than a window inside MTA.

**Comment by Necktrox** [comment]
@qaisjp  I changed the title, but I kept the text for the tray icon tooltip.

Preview

**Comment by qaisjp** [comment]
Nice call! What is the tray icon tooltip at the moment?

**Comment by ccw808** [comment]
My thoughts:
1. When MTA is not minimized, the notifications should be drawn inside MTA. (i.e. with CGraphics functions)
2. Add an option to enable tray notifications when minimized

**Comment by Necktrox** [comment]
I agree with the second point, but I am not sure about the first point, because MTA (not a script) itself could use it to show various notifications and most scripters would rather make their own notification system + visuals than use a non-customizable inbuilt one.

**About the second thought**:
A tray notification, as of now, will only show up when your MTA window is not focused (you can't read them when you are in MTA). And there is a checkbox to enable/disable tray notifications (see the screenshots above).

**Comment by qaisjp** [comment]
I agree with @necktrox

**Comment by ccw808** [comment]
Ok, that's fine then.

**Pull request closed without merging**